### DISCUSSION TITLE
Dynamic block - How to ignore if present but null

### DISCUSSION THREAD

**Discussion by SimonSouli** [aztfmod/terraform-azurerm-caf#1457] [Q&A]
Hello,
I've made a modification on Synapse workspace module in order to get aad_admin, azure devops repo and github repo working.
I now have an issue, as they are not mandatory, but as long as you specify the block (even if null) it's detected as present and waiting for the content.

Below the dynamic block. As you can see if present it will look for content.

```
  dynamic "azure_devops_repo" {
    for_each = try(var.settings.azure_devops_repo, {}) == {} ? [] : [1]

    content {
      account_name    = var.settings.azure_devops_repo.account_name
      branch_name     = var.settings.azure_devops_repo.branch_name
      project_name    = var.settings.azure_devops_repo.project_name
      repository_name = var.settings.azure_devops_repo.repository_name
      root_folder     = var.settings.azure_devops_repo.root_folder
    }
  }
```

One of my try to call it from locals
`azure_devops_repo               = (var.naming.env=="dev" ? local.azure_devops_repo : {})
`
and associated local 
```
locals {
  azure_devops_repo = {
    az_repo1 = {
      account_name    = "${var.synapse_azure_repo.account_name}",
      branch_name     = "${var.synapse_azure_repo.branch_name}",
      project_name    = "${var.synapse_azure_repo.project_name}",
      repository_name = "${var.synapse_azure_repo.repository_name}",
      root_folder     = "${var.synapse_azure_repo.root_folder}"
    }
  }
}
```

Regards,
Simon

If it's not dev, i get below error for each attributes: 
```

 On .terraform\modules\caf\modules\analytics\synapse\workspace.tf line 42, in resource "azurerm_synapse_workspace" "ws":
  42:       project_name    = var.settings.azure_devops_repo.project_name
     ├────────────────
     │ var.settings.azure_devops_repo is null
 This value is null, so it does not have any attributes.
 Error: Attempt to get attribute from null value
```

Any idea on how to modify the condition for the dynamic block in order to allow it ?

```
 dynamic "azure_devops_repo" {
    for_each = try(var.settings.azure_devops_repo, {}) == {} ? [] : [1]
```

**Comment by brk3** [comment]
In your `content` block, try wrapping your arguments in `lookup` functions:

```
  dynamic "azure_devops_repo" {
    for_each = try(var.settings.azure_devops_repo, {}) == {} ? [] : [1]

    content {
      account_name    = lookup(var.settings.azure_devops_repo, "account_name", null)
      branch_name     = lookup(var.settings.azure_devops_repo, "branch_name", null)
      project_name    = lookup(var.settings.azure_devops_repo, "project_name", null)
      repository_name = lookup(var.settings.azure_devops_repo, "repository_name", null)
      root_folder     = lookup(var.settings.azure_devops_repo, "root_folder", null)
    }
  }
```

This way it will allow for a `azure_devops_repo` block containing some/all/none of the settings. Note you still have to comply with what the provider expects, e.g. if the provider docs states that `account_name` must be given then it doesn't make sense to make that optional using `lookup`. Likewise, if the provider docs give a default other than null, you should use that as the last arg to `lookup` rather than null.
